Feature/codeblock in embeddederror in iam,issue,reminder,standup,tips

ChinmayMhatre opened the pull request:
this is a continuation to discussion on #378 added code blocks for the above commands

ChinmayMhatre commented:
![image](https://user-images.githubusercontent.com/51131670/102697571-b8d01080-425c-11eb-8f45-51e2b4c6dfb9.png)

ChinmayMhatre commented:
![image](https://user-images.githubusercontent.com/51131670/102697582-cf766780-425c-11eb-926d-2635b7f0cfd2.png)

ChinmayMhatre commented:
![image](https://user-images.githubusercontent.com/51131670/102697594-e87f1880-425c-11eb-963e-8df218d03096.png)

AllanRegush commented:
I think we should make an error embed function to format all of these commands. Let's keep the code DRY.

ChinmayMhatre commented:
> I think we should make an error embed function to format all of these commands. Let's keep the code

Could you tell me where to put the function ? The function is given in the original issue but I couldn't figure out where to put it

AllanRegush commented:
I think maybe `config.ts`?

ChinmayMhatre commented:
ok I'll try it out

eddiejaoude commented:
Looks good 👍 . 

I think this is almost there, just waiting on the DRY suggestion?

ChinmayMhatre commented:
> Looks good 👍 .
> 
> I think this is almost there, just waiting on the DRY suggestion?

yeah . I'll do it in 1-2 days.

ChinmayMhatre commented:
I think I'll create a new branch and pr for the dry suggestion .

ChinmayMhatre commented:
I am not getting the function part can someone explain it again?

Vyvy-vi commented:
Oh, I'm sorry(the screenshot I attached in the discussion wasn't very clear and explanatory) 😅 

The `function` in there doesn't have much to do(it's a pre-existing function that is used to render default embeds by the bot)
The arrow functions in that ss are the essential part, they do a no. of things-
```js
// Arrow function to convert string into code-block format, with command prefix appended at the start of the string
const mapToCodeBlock = (arg: string) =>
    '`' + `${COMMAND_PREFIX}` + `${arg}`.trim() + '`';
// splits the usage-text into it's separate options
const usageTextOptions = usage.split(' or ')
// formatted usage string that maps all the options to be formatted with code-blocks and then joins them together with ' or ' in between
const formatted = usageTextOptions.map(mapToCodeBlock).join(' or ')
 ```
 Now, this `formated` string can be used in the error embeds instead of `usage` string

Vyvy-vi commented:
We could take the above and put it in a function defined in a file/config file so that it can be used globally in all the command files, which would reduce redundancy

AllanRegush commented:
I think we should use template strings where possible. String concatenation isn't very performant.

Vyvy-vi commented:
I think that was written like that since I didn't know how to use backticks in template strings 😅

ChinmayMhatre commented:
> Oh, I'm sorry(the screenshot I attached in the discussion wasn't very clear and explanatory) 😅
> 
> The `function` in there doesn't have much to do(it's a pre-existing function that is used to render default embeds by the bot)
> The arrow functions in that ss are the essential part, they do a no. of things-
> 
> ```js
> // Arrow function to convert string into code-block format, with command prefix appended at the start of the string
> const mapToCodeBlock = (arg: string) =>
>     '`' + `${COMMAND_PREFIX}` + `${arg}`.trim() + '`';
> // splits the usage-text into it's separate options
> const usageTextOptions = usage.split(' or ')
> // formatted usage string that maps all the options to be formatted with code-blocks and then joins them together with ' or ' in between
> const formatted = usageTextOptions.map(mapToCodeBlock).join(' or ')
> ```
> 
> Now, this `formated` string can be used in the error embeds instead of `usage` string

oh so I just add the arrow functions in the config.ts file?

AllanRegush commented:
@Vyvy-vi was just giving an example on how to create a function to format the embeds. I'd recommend creating a function in `config.ts` using string templates. If you need any help let us know.

ChinmayMhatre commented:
> @Vyvy-vi was just giving an example on how to create a function to format the embeds. I'd recommend creating a function in `config.ts` using string templates. If you need any help let us know.

Yeah , could you tell me what to do . I am not getting exactly what to do.

AllanRegush commented:
> 
> 
> > @Vyvy-vi was just giving an example on how to create a function to format the embeds. I'd recommend creating a function in `config.ts` using string templates. If you need any help let us know.
> 
> Yeah , could you tell me what to do . I am not getting exactly what to do.

Create a function that formats the `usage` embeds. Please put the function in `config.ts`.
Requirements:
1. Needs to format the usage strings.
2. Needs to handle the `or` case.

Sample input: (does not have code block formating)
"${triggers[0]} <add/remove> <role name> or ${triggers[0]} <add/remove> <role name>, <role name>, ..."
Exepected output (has code block formating i.e. wrapped in back ticks \`  )
`${triggers[0]} <add/remove> <role name>` or `${triggers[0]} <add/remove> <role name>, <role name>, ...`

Recipe:
see @Vyvy-vi sample.

If you need help or clarification let us know. 😄

ChinmayMhatre commented:
I think I should create a new branch and pr or this , is that fine?

AllanRegush commented:
> 
> 
> I think I should create a new branch and pr or this , is that fine?

That's fine 👍

Vyvy-vi commented:
maybe this one could be merged, after updation?

Vyvy-vi commented:
> umm, If making the function is the goal should this one be merged at all?

IMHO, this PR does add value

ChinmayMhatre commented:
@Vyvy-vi  in `config.ts` I should create the function under the export defaults right? and then use it before it is passed to the  builderror function ? I

AllanRegush commented:
@ChinmayMhatre Yes, after defaultembed in `config.ts`.
Then use like this. But not exactly this is just a sample.
```ts
  function buildErrorEmbed(errorMsg = 'An error has occurred') {
    return embed
      .setColor(config.COLORS.alerts)
      .setTitle('Role Assignment (error)')
      .setDescription(errorMsg)
      .addField('Usage', myFormatFunc(usage));
  }
 ```

AllanRegush commented:
Just checking in @ChinmayMhatre. If you need any help let us know.

ChinmayMhatre commented:
yeah really sorry ,  had my exams past 2 weeks . Will get back to this tommorow .

ChinmayMhatre commented:
@AllanRegush  Could you tell me how do I make the function in the `config.ts` ? I create a function with the arg that is the usage string that is passed . Then could you explain how to use the string templates to form the end string? do I create the different parts using `spilt(' or ')` like @Vyvy-vi  put in the ss ?

AllanRegush commented:
You can make functions in `config.ts` like this sample
```ts
  },
  defaultEmbed: (color = '#0099ff') => {
    return new MessageEmbed()
      .setColor(color)
      .setTimestamp()
      .setFooter(
        'Our bot is Open Source, you can find it here https://github.com/EddieJaoudeCommunity/EddieBot'
      );
 ```
For template strings here is the documentation: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

ChinmayMhatre commented:
> @ChinmayMhatre Yes, after defaultembed in `config.ts`.
> Then use like this. But not exactly this is just a sample.
> 
> ```ts
>   function buildErrorEmbed(errorMsg = 'An error has occurred') {
>     return embed
>       .setColor(config.COLORS.alerts)
>       .setTitle('Role Assignment (error)')
>       .setDescription(errorMsg)
>       .addField('Usage', myFormatFunc(usage));
>   }
> ```

That part I got . But how do I split the usage variable that comes into the function as the argument?  like in this example you gave it was `myFormatFunc` right . so the usage variable give the whole string along with the ` ' or ' ` condition so how do I split it in the function ?

Vyvy-vi commented:
mini-hint: 
.map(arrow_function)

AllanRegush commented:
use `str.split()` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split

ChinmayMhatre commented:
Could we create a new issue for the DRY implementation and merge this pr for now and let anyone who wants to do it could look at this discussion . I am a little busy with my college projects so I won't be able to give time in solving this issue and this would only get further delay . I feel it would be quicker if another issue was created.

AllanRegush commented:
Thank you for your hard work @ChinmayMhatre. We can close the PR and update the issue

eddiejaoude commented:
Thank you all for your hard work 💪 .

@ChinmayMhatre good luck with your college projects and thanks again 🤓

BOLT04 commented:
> 
> 
> Thank you for your hard work @ChinmayMhatre. We can close the PR and update the issue

Hi @AllanRegush 😃, can we merge this instead of closing, and open a new issue for the DRY implementation?

BOLT04 reviewed (approved):
I think we can merge this since it already adds value 👍, so thank you @ChinmayMhatre 😃.
Let me know what you think @Vyvy-vi and @AllanRegush 👍

Vyvy-vi commented:
I think this can be merged, for now :+1:

eddiejaoude commented:
Thank you all 👍
